serverless chunker

ccraddock opened the pull request:
A few changes to enable running the chunker as a s3-triggered lambda function. Changes include:

- storing raw data in the RAW_DATA folder of the bucket, so that the lambda function can be triggered only on raw data
- updating all relevant file parsing code, etc to accommodate the changes to raw data paths
- writing a new chunker function, that relies heavily on previous function, that can be run as a lambda
- adding scripts to test the functionality

A deploy script for deploying the lambda function is still in progress.

I additionally changed the ChunkRegistry so that file_size refers to number of lines (i.e. observations) for chunkable files, this is a bit easier to understand than bytes. Unchunkable data is still in bytes. The dashboard has not been updated to reflect this change yet.

I have tested the code changes and all appear to be working well. My guess is that this will provide a good cost reduction over the previous celery file processing cluster.

Running the file chunker as an S3-triggered lambda may make some of the open issues no longer relevant specifically: #93, #95, #97

biblicabeebli commented:
oh wow.
I ... will review this.  Very cool.

Wondering if you could comment on server vs lambda costs with this while I review?

Humorously, I need to get my own updated celery task dispatch from development onto master today..... hmmmm... funnnn....

ccraddock commented:
glad you are interested, there is some code that i didn't intend to include in this PR that handles generating key pairs in the chunker. I found a bug in this code and am working on fixing it now. I will submit a new PR once that is done. 

I will also provide some cost comparisons soon.

biblicabeebli commented:
I've done a first pass over this, so far this is fairly solid.  I have not reviewed the big-ole function that you've set up for lambda, I will be going through it.

Lambda was brand new (November 2014, and iirc data processing was built in either December 2014 or January 2015).  It is transparently the perfect fit service for running data processing.  I expect costs will end up lower than a server.  I never did development on it, I'm not that familiar with it except that bundling code together is a bit of a pain.  I am thrilled to see this work you are doing.  I will try and catch up with some of the Lambda documentation, but assume for now I know nothing on the topic.

Rather than try and lump several questions together I will post them as individual comments.

biblicabeebli commented:
Question 1.
How necessary is that RAW_DATA folder?

I've got a deployment with a _**LOT**_ of data in it, moving it is not feasible - well, okay, it _can_ be moved, but we are talking gigabytes just for strings of file names.  I/we/you would have to write a perfect script to track down all database entries containing file paths.  It would be much easier and less error prone for existing deployments to.... "spread out" and manage many triggers.  I think.

biblicabeebli commented:
Question 2
... can we make use of AWS' compute auction services with Lambda?  :D :D

biblicabeebli commented:
Question 3
I guess you've implemented a way to solve #94 using Lambda, cool.

usanchez commented:
are we taking a serverless approach? That would be awesome and I would more than willing to learn more about it to be able to contribute. I don't know which would be the crossing point of the costs of serverless vs elastic beanstalk with relation to the number of participants, but this is indeed an interesting approach indeed.

biblicabeebli commented:
I guess we will find out!  ;D

biblicabeebli commented:
Question 4:
With regards the `except FileToProcess.MultipleObjectsReturned as e:` exception:
This is a bug, I have a dedupe script and a schema migration on development.  I'm dogfooding it, I'm still tracking down the other bugs it uncovers.  (Origin of the error is from the old, crap database (mongo), only got around to deduping this week, its been a problem forever.)

This is related to an important fundamental: the data processing operations are separated by participant, and any given participant can only have one processing task active at a given moment or else interleaved upload operations will are possible.  This results in a silent data loss in the chunked files.  One of the things in the code on development does right now is re-add all potentially relevant source files for every duplicate it finds.  The other side of this is that duplicate ChunkRegistry file paths are the only way I've come up with to detect interleaved writes.  (hm, I suppose testing file sizes of versioned data on s3? maybe?)

As a result I don't know that triggering based on s3 uploads is workable.  I got fed up with the old FileProcessingLock issues too and am transitioning to a better-but-not-perfect timer-based celery solution.  If we can guarantee a good triggering mechanism and safety checks for potential interleaved writes with this Lambda work I will _happily_ ditch celery and manager/worker servers.

biblicabeebli commented:
Question 5:
I see you are using threadpools to run multiple concurrent tasks.  This is not going to help much due to the python GIL, its just going to make errors harder to track down.  Network operations are already threaded in the data processing code.

biblicabeebli commented:
Question 6:
I will try and research this myself as it is an AWS question, but what kind of ram limitations do we have on lambda?  I've seen 500MB+ accelerometer files.  Not common, but it happens.  Data processing is optimized for ram usage more than anything else, but there's no way to avoid at least a 2x memory usage of the size of the chunked file when uploading or downloading. (Really, isn't possible. there's an encryption step, and the hassle of partial writes due to streaming data to S3 is absolutely not worth the problems that will call.)

biblicabeebli commented:
Question 7:
I'm seeing some statements on compression?  Compression is a wishlist todo but I'm not seeing anything like that in your code.  am I missing something?

usanchez commented:
I can answer to question 6:
generally speaking, lambda functions are best suited for methods that don't take a lot of time to be executed and don't take too much memory. Currently, the max memory allocation is 3008 MB and the timeout is set at 15 mins (900 secs). The billing is made depending on the memory used (in increments of 64MB) and the time that it takes to run the function.

More info on: https://docs.aws.amazon.com/lambda/latest/dg/limits.html

biblicabeebli commented:
Q6:
Thanks Unai.  I just created an enhancement issue for some of this possible streaming optimization.
Currently long-running data processing is a possibility if a user is constantly uploading data while the processing is occurring.  This detail can be changed.  There is also the batch size parameter, which is set to 250 based purely on an "I had to stare at cpu and network usage graphs" at some point to see how long data processing took. (Also, a larger file-batch size is more efficient with respect to S3 operations.)
--Point is, the 15 minute timeout is solvable.... as long as we don't get interleaved s3 writes.

usanchez commented:
Regarding my previous post about the serverless approach, I was thinking that it could be interesting to think about the possibilities that the following AWS services may offer:
- **AWS Cognito**: User credentials, keys and so on. This could ease encryption tasks.
- **AWS DynamoDB**: the RDS is responsible for a big part of the billing. Implementing DynamoDB would be awesome, but is not compatible with Django as a Django database. There is some programming effort to make the Django models fit the DynamoDB structure. We would lose the "migrations" too.

There are more and more complex use cases, these are just the ones in the top of my head that may be interesting to consider.

ccraddock commented:
> Question 1.
> How necessary is that RAW_DATA folder?
> 
> I've got a deployment with a _**LOT**_ of data in it, moving it is not feasible - well, okay, it _can_ be moved, but we are talking gigabytes just for strings of file names. I/we/you would have to write a perfect script to track down all database entries containing file paths. It would be much easier and less error prone for existing deployments to.... "spread out" and manage many triggers. I think.

As a caveat, I am building a new deployment of Beiwe and am not concerning myself with legacy data. You are right that does create a pain.

This is an issue I have dealt with quite a bit. The major concern is the file that you write might once again trigger the lambda. Unfortunately s3 triggers for lambda aren't extremely configurable. You cannot configure the trigger to ignore data written to a specific folder. You could have your lambda function abort if it finds out that the data is from the 'CHUNKED_DATA' folder, but that is a waste of resources.

To avoid changing the file path, you could register a separate trigger for each study directory, which is what I think you are alluding to in your question. This could theoretically be done as a part of the 'create new study' api. This seems like a more complicated option so I didn't go down that path. But it is entirely possible.

If all of your legacy data has already been chunked, it is also possible to just leave it where it is and only use the serverless chunker for new data. I did this with my python2 installation of beiwe. You just need to make sure that all of the filepath parsing can handle it. If you are interested, you can see how i handled it in the python2 version here: https://github.com/ccraddock/beiwe-backend/tree/wcwh-mods-py2

ccraddock commented:
> FileToProcess.MultipleObjectsReturned

As I understand it, there are two problems that you are referring to. This first is that for some reason an upload can be added twice to the FilesToProcess model, the other is that you may receive two files that each contain data for the same time bin. I haven't spent a ton of time looking through how the first problem happens, but since the filenames are the same, both can only refer to the same file, if there were two different files with the same filename one of them has been written over. So I just used the latter. The second problem, as you mention is a humdinger.

One nice advantage of lambda is that essentially you are running a different process for each new file. If you receive 1000 files at the exact same time, the system could spin up 1000 seperate resources to execute the chunker on each of the files in parallel. Practically, not all of the files are received at the exact same time, so you will not get that many running at once. It also doesn't necessarily spin up new resources for each file, if a file comes in near the completion of the processing for a previously received file, then those resources will be used to process the second file. This is to reduce the overhead of spinning up resources to handle the files. 

Using this model, if each 'raw data' file contains completely independent data then there is no concern. It will generate independent chunked files, and there is no need for mutexs or other precautions. The only concern might be with the database, but that should have internal mechanisms for handling things. But as you point out, and I didn't fully appreciate before, if two files are received at the same time and processed by different lambdas, then one could theoretically overwrite the other. I wonder if S3 has a process to avoid this from happening?

I will do more research into this and try to figure it out. One solution that comes to mind is that we can add a random string to the filename of each chunked file. This would avoid the problem of overwriting the file on s3. Since the ChunkRegistry doesn't require any of the fields to be unique (i see that was removed) then we don't need to worry about collisions there. Instead, the pipeline developers would need to be aware that there could be multiple chunk files for a time bin and code it accordingly. Is that a reasonable solution?

ccraddock commented:
> Question 5:
> I see you are using threadpools to run multiple concurrent tasks. This is not going to help much due to the python GIL, its just going to make errors harder to track down. Network operations are already threaded in the data processing code.

Are you referring to the ThreadPool in [upload_binified_data](https://github.com/ccraddock/beiwe-backend/blob/16161900d745ef7f14a21d6c96e1e24ae4ee50ff/libs/file_processing.py#L483)? That is legacy code that I failed to change. I have a new PR in the works that will address this issue.

ccraddock commented:
> Question 6:
> I will try and research this myself as it is an AWS question, but what kind of ram limitations do we have on lambda? I've seen 500MB+ accelerometer files. Not common, but it happens. Data processing is optimized for ram usage more than anything else, but there's no way to avoid at least a 2x memory usage of the size of the chunked file when uploading or downloading. (Really, isn't possible. there's an encryption step, and the hassle of partial writes due to streaming data to S3 is absolutely not worth the problems that will call.)

Each file is processed by a seperate lambda run, so the resources that are required are just what is needed by a single file. Of course, this will depend on the app settings. I imagine 500MB+ acceleromter files used pretty aggressive sampling. As mentioned by @usanchez the limits are 3GB and 5 minutes of processing. The amount of CPU time allocated to your lambda is scaled by the amount of memory you allocate. You pay based on a time*memory. The time is what you use, the memory is what is allocated, not what is used. In my tests, with smaller accelerometer files, I have been able to get by with 192MB and each file completes in well under a second.

ccraddock commented:
> Question 7:
> I'm seeing some statements on compression? Compression is a wishlist todo but I'm not seeing anything like that in your code. am I missing something?

The file contents are compressed before sending to the code that uploads the [see here](https://github.com/ccraddock/beiwe-backend/blob/16161900d745ef7f14a21d6c96e1e24ae4ee50ff/libs/file_processing.py#L448). It is then [uncompressed before being uploaded](https://github.com/ccraddock/beiwe-backend/blob/16161900d745ef7f14a21d6c96e1e24ae4ee50ff/libs/file_processing.py#L848), but the compressed version is sent to [ChunkRegistery.register_chunked_data](https://github.com/ccraddock/beiwe-backend/blob/16161900d745ef7f14a21d6c96e1e24ae4ee50ff/libs/file_processing.py#L865). In the code I added I had to uncompress the data in register_chunked_data so that I could calculate the number of rows in the file, [here](https://github.com/ccraddock/beiwe-backend/blob/16161900d745ef7f14a21d6c96e1e24ae4ee50ff/database/data_access_models.py#L93).

ccraddock commented:
> Q6:
> Thanks Unai. I just created an enhancement issue for some of this possible streaming optimization.
> Currently long-running data processing is a possibility if a user is constantly uploading data while the processing is occurring. This detail can be changed. There is also the batch size parameter, which is set to 250 based purely on an "I had to stare at cpu and network usage graphs" at some point to see how long data processing took. (Also, a larger file-batch size is more efficient with respect to S3 operations.)
> --Point is, the 15 minute timeout is solvable.... as long as we don't get interleaved s3 writes.

Since each file is processed in a different lambda instance, the timeout only applies to processing a single file, do you think we will run into danger of that happening?

ccraddock commented:
By the way, this PR is primarily concerned with serverless chunking, but I am also very interested in running the entire backend serverless using [API Gateway](https://aws.amazon.com/api-gateway/). It works! 

The real advantage is that it will automagically scale to allow up to 1000 simultaneous (as in exactly simultaneous) api calls. My hope is that this will scale much better to very large studies. The potential cost savings are just gravy. With EBS and elastic load balancing you get a good deal of expandability, but with API Gateway, you get this scalability on the millisecond scale.

Also with [zappa](https://github.com/Miserlou/Zappa) you can do this with FLASK apps (like beiwe) with almost no work! And then it works just like EBS, git pull, zappa update, and you are golden.

I haven't added the config files to the repos yet because I am paranoid about revealing account details, but I would be happy to talk more about this.

biblicabeebli commented:
I typed up a lot of stuff on a commute, figure I'll paste here and get more done later.


> Since each file is processed in a different lambda instance, the timeout only applies to processing a single file, do you think we will run into danger of that happening?
I do not think timing out will be an issue.

> Since the ChunkRegistry doesn't require any of the fields to be unique (i see that was removed) then we don't need to worry about collisions there.
The lack of uniqueness was not removed, it was a bug that was only discovered when we moved off an old database technology.  Work currently on the development branch includes a script to fix this and adds database enforcement.  The reason I am doing this is so that chunking catches the issue (it hits the database constraint) and fails without retiring any FileToProcess entries. The next periodic data processing task creation event (under the current celery data processing paradigm) will operate on any and all unprocessed data that remains, solving the s3 write interleave problem.  The current (master) logic uses a database lock, but apparently some write interleaves occured anyway as evidenced by the presence of duplicate file paths in the chunk registry on more than one deployment I manage.

> [unique filename strings idea] ... Instead, the pipeline developers would need to be aware that there could be multiple chunk files for a time bin and code it accordingly. Is that a reasonable solution?
No. The detail you describe is perceived as a problem by data scientists (plus fixing out-of-order observations, not uncommon, and making the data queryable by time) that chunking exists in the first place.  My invariant experience has been that anything more complex than "You'll get your data, just check in periodically until the participant uninstalls the Beiwe app, there is an api and reference python script if you want to automate this." is totally unworkable.  My necessary conclusion is that few data scientists have sufficient expertise to automate piecing data together, and for that reason we have the data access api and must clean up the data for them.

Other options I've considered:
At-upload data processing - doesn't solve anything.
Per-file locks - files can overlap an hour boundary, resulting in potential for write interleaves
Per-file locks, but separate out the hourly chunks as a separate step - introduces another write interleave step.
Use a database - petabytes of data.  Formatting could be made binary, but that's still terabytes of data.
Do it on the device - the device data encryption is implemented using a public-private key pairing. The device cannot decrypt what it writes. This paradigm could be totally ditched, I acknowledge it is a theoretically workable option, but I am not the person to develop it.

(I am _very happy_ you are thinking about this, I do not have a great answer and I wish to encourage you on you work.  This is a problem and I have a love-hate relationship with Celery too.)


> I wonder if S3 has a process to avoid this from happening?
Pretty sure that's a nope.  As far as I know partial writes and appends are not an option, neither are any atomic guarantees.  We use s3 because it has decent data integrity and is relatively cheap.  File versioning could assist with interleaved write detection, but at some point there has to be a single file containing all the data in a time bin.


> DynamoDB
We moved away from a nosql database solution (MongoDB) because it caused problems, was eggregiously slow once data approached large, and required I manage an extra server with 10x the storage needs.  It also crashed.  Ever.  My personal opinion is that the advantage of alternative, non-relational database technologies is either negligible or negative.  Yes, RDS is expensive, but the solution to the cost problem is to add an option to the deploy script to set up a custom non-RDS database server.  It can be mysql, mariadb, postgres, or even sqlite.  Any mature relational database technology.  We will not be moving away from the current relational database paradigm.


Some additional commentary based on 5 years of this:
- researchers _love_ that data is available quickly.
- researchers _usually_ understand that data comes in out-of-order and are okay with it.
- researchers _nearly universally_ prefer out-of-order-but-quick is preferable in-order-but-slow.  When they are hesitant about it they change their mind as soon as they see that they have partial data from all the data streams and can start playing with it immediately.
- researchers will tend to notice within a day (weekday) when data processing is down, especially once they catch on to the fact that most data is uploaded at night while the phone is plugged in, on wifi, and charging.

- meta commentary: incomplete data is a fundamental fact that researchers already have to deal with. Participants turn off their phone, enable low power mode, run out of storage, forget to turn on wifi, run out of battery, etc. Their end-data _will_ be inconsistent. The fact that we happen to mimick this in data upload sets up reasonable expectations.  Happy to find ways to improve this and have spent a lot of effort doing so, it is not a fully resolvable problem.

biblicabeebli commented:
Zappa:
This does look really cool.
Um, I'm slightly concerned that the second-to-top issue on the github is asking if it is still being developed with an answer stating effectively "no"?

I'm not familiar, you appear to be, happy to think about it.  FYI though, while this project is open source we do have active clusters and clients paying for system administration and feature development.  One of the features we just finalized involves push notifications to devices, and that sounds like something that would immediately conflict with "serverless."

ccraddock commented:
Hello Eli,

Thanks for the detailed comment. I fit into the developer, dev/ops and
researcher categories. My experience is highly consistent with yours in
terms of what others expect. But I think we may be able to address some of
the issues you describe with abstraction, esp if the data is
primarily accessed through the API.

To make sure that I am clear, I am assuming that the interleaved write
issue that you are referring to occurs when observations (rows of a data
file) corresponding to a specific time bin are spread across two (or more?)
upload files. This creates a 'multiple writer' problem because the file
processing (be it celery or lambda -based) for the two separate files may
occur simultaneously by two separate threads or processes that do not know
anything about each other. For S3 the resulting collision may result in the
data from one of the upload files (partial time bin data) to overwrite the
data from the second upload. If the ChunkRegistry requires that the
combination of timestamp and data type to be unique the same problem could
occur in the database (it doesn't now so no problem).

You have thought through quite a few different solutions, that I agree are
all unattractive for one reason or the other, esp due to their complexity.
My approach to solving this issues is usually to remove the dependencies,
rather than trying to use mutexs or the like.  Here is the solution I am
working on now, which I think will solve the problem for my needs, but may
not be preferred by your clientele:

Instead of naming the chunk files based on their time bin
(e.g. 2020-01-05T02:00:00.csv), name the file based on the time-stamp of
the first observation in the file. In this
case 2020-01-05T02:17:32.007.csv. I believe that this will avoid collisions
on S3, unless multiple observations are acquired at the same millisecond,
(if that does happen I think I will have bigger problems to worry about).
Even in the case that the partial files are perfectly interleaved (i.e. odd
observations in file 1 and even in file 2) this should cause a problem
because the first time stamp should still be different.

For the database, I will keep the database as is, and allow multiple files
per chunk for each data stream. For my customers, this doesn't cause a
problem because they are just going to concatenate all of the files that
they download anyway. But if they do have a problem with this, I will
likely modify the data download api to reformat and concatenate so that it
looks like it did before.

I agree that the lack of active maintenance of Zappa is discouraging. I
referenced it because it makes it amazingly easy to deploy beiwe
serverless. I am leaning towards terraform (https://www.terraform.io/) for
a long term solution, I commend you and your colleagues on the AWS deploy
scripts you have written, but I am hesitant to develop similar ones for the
serverless deploy.

Cheers,
Cameron


On Fri, Jan 10, 2020 at 6:54 PM Eli Jones <notifications@github.com> wrote:

> I typed up a lot of stuff on a commute, figure I'll paste here and get
> more done later.
>
> Since each file is processed in a different lambda instance, the timeout
> only applies to processing a single file, do you think we will run into
> danger of that happening?
> I do not think timing out will be an issue.
>
> Since the ChunkRegistry doesn't require any of the fields to be unique (i
> see that was removed) then we don't need to worry about collisions there.
> The lack of uniqueness was not removed, it was a bug that was only
> discovered when we moved off an old database technology. Work currently on
> the development branch includes a script to fix this and adds database
> enforcement. The reason I am doing this is so that chunking catches the
> issue (it hits the database constraint) and fails without retiring any
> FileToProcess entries. The next periodic data processing task creation
> event (under the current celery data processing paradigm) will operate on
> any and all unprocessed data that remains, solving the s3 write interleave
> problem. The current (master) logic uses a database lock, but apparently
> some write interleaves occured anyway as evidenced by the presence of
> duplicate file paths in the chunk registry on more than one deployment I
> manage.
>
> [unique filename strings idea] ... Instead, the pipeline developers would
> need to be aware that there could be multiple chunk files for a time bin
> and code it accordingly. Is that a reasonable solution?
> No. The detail you describe is perceived as a problem by data scientists
> (plus fixing out-of-order observations, not uncommon, and making the data
> queryable by time) that chunking exists in the first place. My invariant
> experience has been that anything more complex than "You'll get your data,
> just check in periodically until the participant uninstalls the Beiwe app,
> there is an api and reference python script if you want to automate this."
> is totally unworkable. My necessary conclusion is that few data scientists
> have sufficient expertise to automate piecing data together, and for that
> reason we have the data access api and must clean up the data for them.
>
> Other options I've considered:
> At-upload data processing - doesn't solve anything.
> Per-file locks - files can overlap an hour boundary, resulting in
> potential for write interleaves
> Per-file locks, but separate out the hourly chunks as a separate step -
> introduces another write interleave step.
> Use a database - petabytes of data. Formatting could be made binary, but
> that's still terabytes of data.
> Do it on the device - the device data encryption is implemented using a
> public-private key pairing. The device cannot decrypt what it writes. This
> paradigm could be totally ditched, I acknowledge it is a theoretically
> workable option, but I am not the person to develop it.
>
> (I am *very happy* you are thinking about this, I do not have a great
> answer and I wish to encourage you on you work. This is a problem and I
> have a love-hate relationship with Celery too.)
>
> I wonder if S3 has a process to avoid this from happening?
> Pretty sure that's a nope. As far as I know partial writes and appends are
> not an option, neither are any atomic guarantees. We use s3 because it has
> decent data integrity and is relatively cheap. File versioning could assist
> with interleaved write detection, but at some point there has to be a
> single file containing all the data in a time bin.
>
> DynamoDB
> We moved away from a nosql database solution (MongoDB) because it caused
> problems, was eggregiously slow once data approached large, and required I
> manage an extra server with 10x the storage needs. It also crashed. Ever.
> My personal opinion is that the advantage of alternative, non-relational
> database technologies is either negligible or negative. Yes, RDS is
> expensive, but the solution to the cost problem is to add an option to the
> deploy script to set up a custom non-RDS database server. It can be mysql,
> mariadb, postgres, or even sqlite. Any mature relational database
> technology. We will not be moving away from the current relational database
> paradigm.
>
> Some additional commentary based on 5 years of this:
>
>    -
>
>    researchers *love* that data is available quickly.
>    -
>
>    researchers *usually* understand that data comes in out-of-order and
>    are okay with it.
>    -
>
>    researchers *nearly universally* prefer out-of-order-but-quick is
>    preferable in-order-but-slow. When they are hesitant about it they change
>    their mind as soon as they see that they have partial data from all the
>    data streams and can start playing with it immediately.
>    -
>
>    researchers will tend to notice within a day (weekday) when data
>    processing is down, especially once they catch on to the fact that most
>    data is uploaded at night while the phone is plugged in, on wifi, and
>    charging.
>    -
>
>    meta commentary: incomplete data is a fundamental fact that
>    researchers already have to deal with. Participants turn off their phone,
>    enable low power mode, run out of storage, forget to turn on wifi, run out
>    of battery, etc. Their end-data *will* be inconsistent. The fact that
>    we happen to mimick this in data upload sets up reasonable expectations.
>    Happy to find ways to improve this and have spent a lot of effort doing so,
>    it is not a fully resolvable problem.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/onnela-lab/beiwe-backend/pull/102?email_source=notifications&email_token=AANFGWGLNPAXKYPMZBXEIK3Q5DU7ZA5CNFSM4KEXBBVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIVKJCQ#issuecomment-573219978>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANFGWC7KKVAF2SYWF2EBATQ5DU7ZANCNFSM4KEXBBVA>
> .
>


-- 
Associate Professor of Diagnostic Medicine, Dell Medical School at the
University of Texas at Austin
Director, Computational NeuroImaging Lab
<http://computational-neuroimaging-lab.org/>
Founding Chair, Organization for Human Brain Mapping Open Science Special
Interest Group
Co-Founder, The Neuro Bureau <http://neurobureau.org/>
Impact Story <https://impactstory.org/CameronCraddock>
Google Scholar Citations
<http://scholar.google.com/citations?user=lZ9hwSsAAAAJ&hl=en>
GitHub <https://github.com/ccraddock>
SlideShare <http://www.slideshare.net/CameronCraddock>

Check out my blog Open Science Resources for Contributing to Brain Imaging
Research
<http://blogs.biomedcentral.com/gigablog/2016/03/15/open-science-resources-contributing-brain-imaging-research-guest-blog-cameron-craddock/>
Share that Brain @ the International Neuroimaging Datasharing Initiative
<http://fcon_1000.projects.nitrc.org/>!
The GigaScience Brainhack Thematic Series
<http://brainhack.org/brainhack-thematic-series-call-for-papers/> is open
for your submission.
Submit a paper to the Neuroimage Special Issue on Brain Segmentation and
Parcellation
<http://www.journals.elsevier.com/neuroimage/call-for-papers/brain-segmentation-and-parcellation/>
!
The Preprocessed Connectomes Project
<http://preprocessed-connectomes-project.org/> is now sharing preprocessed
versions of the ABIDE and ADHD200 neuroimaging datasets.

Join me at upcoming Brainhack <http://brainhack.org/> Events:
2019.07.11 - 2019.13.11 Brainhack Global
<http://www.brainhack.org/global2019>

ccraddock commented:
Hey Eli,

My guess is that you could do the push notifications with lambda. AWS has a
variety of different serverless technologies that can be used in
conjunction to solve a variety of different problems. There are several
different ways to invoke a lambda including using a timer (
https://docs.aws.amazon.com/lambda/latest/dg/lambda-invocation.html).
Amazon step function allow you to link many different lamba functions to
implement state machines (https://aws.amazon.com/step-functions/) among
other things.

Thanks for telling me more about the Beiwe business model. When a client
pays for feature development, are those features pushed into the open
source version? Will push notifications be made available in the open
source version?

Cheers,
Cameron

On Fri, Jan 10, 2020 at 7:03 PM Eli Jones <notifications@github.com> wrote:

> Zappa:
> This does look really cool.
> Um, I'm slightly concerned that the second-to-top issue on the github is
> asking if it is still being developed with an answer stating effectively
> "no"?
>
> I'm not familiar, you appear to be, happy to think about it. FYI though,
> while this project is open source we do have active clusters and clients
> paying for system administration and feature development. One of the
> features we just finalized involves push notifications to devices, and that
> sounds like something that would immediately conflict with "serverless."
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/onnela-lab/beiwe-backend/pull/102?email_source=notifications&email_token=AANFGWBAFMCL23DOY2ONJMLQ5DWBTA5CNFSM4KEXBBVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEIVK7ZY#issuecomment-573222887>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANFGWCZVIYORHTGY5I22FDQ5DWBTANCNFSM4KEXBBVA>
> .
>


-- 
Associate Professor of Diagnostic Medicine, Dell Medical School at the
University of Texas at Austin
Director, Computational NeuroImaging Lab
<http://computational-neuroimaging-lab.org/>
Founding Chair, Organization for Human Brain Mapping Open Science Special
Interest Group
Co-Founder, The Neuro Bureau <http://neurobureau.org/>
Impact Story <https://impactstory.org/CameronCraddock>
Google Scholar Citations
<http://scholar.google.com/citations?user=lZ9hwSsAAAAJ&hl=en>
GitHub <https://github.com/ccraddock>
SlideShare <http://www.slideshare.net/CameronCraddock>

Check out my blog Open Science Resources for Contributing to Brain Imaging
Research
<http://blogs.biomedcentral.com/gigablog/2016/03/15/open-science-resources-contributing-brain-imaging-research-guest-blog-cameron-craddock/>
Share that Brain @ the International Neuroimaging Datasharing Initiative
<http://fcon_1000.projects.nitrc.org/>!
The GigaScience Brainhack Thematic Series
<http://brainhack.org/brainhack-thematic-series-call-for-papers/> is open
for your submission.
Submit a paper to the Neuroimage Special Issue on Brain Segmentation and
Parcellation
<http://www.journals.elsevier.com/neuroimage/call-for-papers/brain-segmentation-and-parcellation/>
!
The Preprocessed Connectomes Project
<http://preprocessed-connectomes-project.org/> is now sharing preprocessed
versions of the ABIDE and ADHD200 neuroimaging datasets.

Join me at upcoming Brainhack <http://brainhack.org/> Events:
2019.07.11 - 2019.13.11 Brainhack Global
<http://www.brainhack.org/global2019>

biblicabeebli commented:
Hey, I've seen this come in, I don't have the time at the moment to go through it immediately, I will try to get to it Monday.  Unfortunately my availability after that is going to be limited because I am going on a much delayed vacation... followed by a jury duty summons and possible jury duty... followed by unrelated work commitments that will inevitably vie for my time over this... followed by the work for onnela lab that I posted an issue if yesterday (though I'm hoping to wrangle a coworker for some of that).

щ(ºДºщ)

... maybe a phone call would help speed this up?  I'll hit you up out of bound.

biblicabeebli commented:
The fast one:

Beiwe is open source, it is under a pretty permissive license.  I plan to have all work I do on the backend repo be publically accessble and documented, work by others may not be and doesn't have to be.  The Apps aren't on as tight a leash, but development on them generally has to be coordinated with the backend repo pretty closely.  I want to be as supportive as possible of the community upstreaming development back to the onnela lab repositories and I will be / am making an effort to incorporate community additions.  I would love to see what you are working on work out, but I also have legacy systems to administrate.  Breaking changes need to be handled gracefully, so I've got a wiki page that will be updated blog-style whenever something like that occurs.

The work with push notifications will  be open source and public, part of the primary repos for the backend, the iOS app, and the Android App.  I'm going to try and make these a non-required feature, but development on it has not started yet so I don't know 100%.  Backend will try to avoid new required parameters and credentials, apps will try to avoid a new required permissions.  This feature is not currently intended to be mandatory, but changes to the mobile OSes may make them functionally necessary.)

I cannot speak to work done that relates to analysis code, that would be under whatever academic (or proprietary/commercial) restrictions the author-researcher is subject to.

biblicabeebli commented:
It took me a couple reads but you've got it correct.  At first I thought you didn't, here's what I wrote up when I thought that, it may be useful, I also describe the strategy I am currently testing.

> To make sure that I am clear, I am assuming that the interleaved write issue that you are referring to occurs when observations (rows of a data file) corresponding to a specific time bin are spread across two (or more?) upload files. This creates a 'multiple writer' problem because the file processing (be it celery or lambda -based) for the two separate files may occur simultaneously by two separate threads or processes that do not know anything about each other. For S3 the resulting collision may result in the data from one of the upload files (partial time bin data) to overwrite the data from the second upload. If the ChunkRegistry requires that the combination of timestamp and data type to be unique the same problem could occur in the database (it doesn't now so no problem).

The chunking process takes many input files and sticks them into their hourly timestamped files.  It does the following:
1) it cleans up and bins/chunks the newly uploaded data into hourly pieces.
2) it pulls in the existing already-chunked-data and merges the new data into it.
3) it deduplicates any duplicate lines (this makes the merging process idempotent, it allows re-adding files if something weird happened, generally is nice to have.)
4) then it writes that merged data out as the new version of the hourly file on s3.

What I mean by interleaved writes occus when:
1) there are 2 (or more) concurrent processes for user X
2) each process pulls a different set of new files but makes use of the same already-chunked-data.
3) At that point, after merging new data with the already-chunked-data each process has created unique content.
4) each process writes to s3 (and updates the database's hash), but only the chronologically last operated-on-data-set is present on s3.  No data from other merge operations is accessible.
5) all tasks have now run to completion, there were no errors, but we had silent data loss.

-- right now I'm seeing there is a chance in there to re-check the database's hash against the data on s3. I think that is another minimization of the issue and not an actual fix. --

There some things that mitigate (but don't actually solve for this) in the current Celery strategy:
Files are batched together.  At the moment it is simply a number of concurrent files, 250.  This considerably reduces the exposure to the bad conditions because it makes writes and reads of already-chunked-data much less frequent.

More critically we also lock on the prior data processing run make sure that we don't have multiple concurrent processing events.

This has been rearchitected on the development branch:

Jobs are queued up per-user-with-unprocessed data, user jobs are not created if that user has an active job currently running.  All jobs expire every 5 minutes and 30 seconds, jobs are only generated every 6 minutes  (Operational details: jobs don't stop running if they take more time this time, they run until they are finished. Jobs have an absolute expiry, not a 5.5 minute lifetime.)  So far this has resulted in better resource utilization and no overlaps or interleaved write issues (the database running these has the column set to unique.)

biblicabeebli commented:
> "If the ChunkRegistry requires that the combination of timestamp and data type to be unique the same problem could occur in the database (it doesn't now so no problem)."

I disagree, this is a problem.  The column is supposed to be unique, data processing can crash when it encounters non-unique objects in that column.  This also causes occasional weirdnesses with zip file generation during data download, no crashes though.

biblicabeebli commented:
> Instead of naming the chunk files based on their time bin (e.g. 2020-01-05T02:00:00.csv), name the file based on the time-stamp of the first observation in the file. In this case 2020-01-05T02:17:32.007.csv. I believe that this will avoid collisions on S3, unless multiple observations are acquired at the same millisecond, (if that does happen I think I will have bigger problems to worry about). Even in the case that the partial files are perfectly interleaved (i.e. odd observations in file 1 and even in file 2) this should cause a problem because the first time stamp should still be different.

Hokay, there is a solid thought here, it could be made to work.

Unfortunately... multiple observations can be acquired on some data streams (accelerometer, surveytimings theoretically, wifi is weird, bluetooth theoretically but it has its own weirdnesses) at the same millisecond. Fortunately the chance of that occurring _across multiple files_ requires sub-millisecond multiple file creation, key generation, data encryption, and write operations on the device, so .... really really shouldn't happen.  Also the files would overwrite each other I think.

There are also some weird situations where duplicate uploads occur, but that is not really an issue.

> For the database, I will keep the database as is, and allow multiple files per chunk for each data stream. For my customers, this doesn't cause a problem because they are just going to concatenate all of the files that they download anyway. But if they do have a problem with this, I will likely modify the data download api to reformat and concatenate so that it looks like it did before.

I as thinking down this line reading what you were saying, my personal reaction is "oh gawd gross..." but that doesn't mean it wouldn't work.

Here's my ask as you develop: update me periodically, and implement it with a different url target.  I specifically wrote the data access api to have version number (v1), please write yours with a target of v2.

My wishlist for data download is to also have the data be compressed as zip files when uploaded to s3 (many benefits there, compute is cheap), and to stream out an uncompressed zip of compressed zips.  I've never run this by a data scientist.

Warning: do not use deflate as a compression mechanism while streaming data, it requires a big dictionary be kept around in memory and you will get memory errors.  Gzip shouldn't do that, but I haven't tested it.  Live compression was a thing in here until the memory errors started happening, gzip was not available in python 2.

biblicabeebli commented:
> I agree that the lack of active maintenance of Zappa is discouraging. I referenced it because it makes it amazingly easy to deploy beiwe serverless. I am leaning towards terraform (https://www.terraform.io/) for a long term solution, I commend you and your colleagues on the AWS deploy scripts you have written, but I am hesitant to develop similar ones for the serverless deploy.

I've heard from several sources both related to and unrelated to Beiwe about terraform.  Please keep me updated I guess, I understand it can handle non-AWS IAAS?

ccraddock commented:
> > I agree that the lack of active maintenance of Zappa is discouraging. I referenced it because it makes it amazingly easy to deploy beiwe serverless. I am leaning towards terraform (https://www.terraform.io/) for a long term solution, I commend you and your colleagues on the AWS deploy scripts you have written, but I am hesitant to develop similar ones for the serverless deploy.
> 
> I've heard from several sources both related to and unrelated to Beiwe about terraform. Please keep me updated I guess, I understand it can handle non-AWS IAAS?

yes terraform can handle IAAS on many different platforms: https://www.terraform.io/

![image](https://user-images.githubusercontent.com/1725272/72371434-22f24400-36e3-11ea-8da3-0954bb56b233.png)

ccraddock commented:
I am closing this PR as I think we have decided that I will keep the serverless development separated for now and we can re-touch base in the future. I will do my best to keep my fork up to date with the latest development on this repo. I will also submit non-serverless related PRs as they are relevant

biblicabeebli commented:
Okay, happy to talk more if parts of the codebase give you trouble.

I just finished squashing the new expression of a duplicate upload bug that was causing issues in data processing.  It currently just uses the newer file as the upload behavior is to overwrite.

Let me know if you take a different approach, I'm adding a todo about detecting that case at upload and handling it differently.  (this issue also occurs if you re-add unchunkable files for processing using UploadTracking, so the fix from file processing should be incorporated.

Also, I'm hoping to have some time to refactor file processing code? its not super high priority though.
